### ISSUE TITLE
Peak unified RAM on 7B parallel bench: Metal buffer cache not released between cases

### ISSUE THREAD

**Issue by bnsd55** [bnsd55/jevmlx#79]
## Summary

On an M5 Max (128 GB) a **single** `mlx-community/Qwen2.5-7B-Instruct-4bit` `jevmlx bench` run drives Activity Monitor to ~100% unified RAM (swap in use), then later reports ~95% free. Process RSS stays ~188 MB. This is not a second model and not “7B × N questions.” It is peak GPU working set (logits slab) plus the Metal **buffer cache**, which is not cleared between cases.

Same 7B used for coding / chat / API stays in a much smaller band because those paths score **one next token**, not `rows × tokens × vocab`.

Observed during the M5 freeze smoke ([#63](https://github.com/bnsd55/jevmlx/issues/63)), commit `5e175a8`. **Not patched** (runbook: tolerances/engine frozen).

## Environment

| | |
|---|---|
| Commit | `5e175a8` (M5 freeze) |
| Chip | Apple M5 Max, 128 GB unified |
| macOS | 26.6.2 |
| mlx / mlx-lm | 0.32.2 / 0.31.3 |
| Command | `.venv/bin/python -m benchmarks.m5 --out m5-2026-09-19 --parity-models mlx-community/Qwen2.5-7B-Instruct-4bit` |
| Live process | one `jevmlx bench --model quality` (pid 53148). No other MLX process. |

Doctor: AC power, editable install → this checkout, quality 7B already in HF cache.

## What we measured

- **One Python process.** RSS ≈ 188 MB the whole time. VSZ huge (GPU maps).
- Activity Monitor / `memory_pressure`: Pages free briefly ~7k (~110 MB) and swap used (~2 GB of 3 GB). Later: “System-wide memory free percentage: 95%.” Spike, then drop — not a monotonic leak of *answers*.
- Log while scoring bundled, then typesafe:

```
Chunking heuristic: 28 rows over 28 passes (rows_per_chunk=1)
parity: FAIL (winners_identical=True, max drift 0.051354 nats, atol 0.05)
```

Parity fail-closed on 7B is expected (#63). Winners identical. Not the RAM issue.

- Typesafe combo (`parallel-slots-typesafe`, `permutations=rotations`) still in progress: 100+ `completed_cases`, `predictions.jsonl` growing. Contexts in the JSONL: 45 records, char length min/p50/max = 0 / 1381 / 34166.

## Why chat/API RAM is lower (same 7B)

| | Chat / coding / generate API | `jevmlx` parallel bench |
|---|---|---|
| Batch | 1 sequence | many rows (field / option / trie branch) |
| Logits | ~`1 × 1 × vocab` per decode step | ~`rows × suffix_tokens × vocab` float32 per chunk |
| After the step | small tensor dies | Metal **keeps** the freed GPU buffers |
| Extra load | one prompt | typesafe **+ cyclic rotations** of every enum |

Vocab for Qwen2.5 is ~152k. Engine comment (accurate): float32 logits are the dominant activation.

```
# jevmlx/engine.py  (run_parallel_generation)
# The estimate includes the [rows, width, vocab] output logits for one chunk
# (float32 logits are the dominant activation). This is a chunking heuristic,
# not a hard bound on peak Metal memory.
```

Chat never builds that slab. The 7B weights (~4 GB 4-bit) are the same. The **job** is not.

## What is already dropped vs what is kept

```
kept for the whole 7B bench:   model weights (load_engine lru, maxsize=1)
new each decide() call:        make_prompt_cache() KV
dropped when decide() returns: that KV (Python ref gone)
NOT cleared per case:          mx.metal buffer cache
cleared after ALL combos:      clear_engine_cache() in run_bench finally
cleared between models only:   mx.metal.clear_cache() in run_bench_models
```

**Single-model** `jevmlx bench --model quality` goes through `run_bench`, not `run_bench_models`. So during this smoke, **`mx.metal.clear_cache()` is never called** until process exit.

```python
# jevmlx/bench.py — run_bench finally (after every combo, once)
clear_engine_cache()          # drops the Engine / weights cache
# does NOT call _clear_metal_cache()

# jevmlx/bench.py — run_bench_models finally (per model)
clear_engine_cache()
_clear_metal_cache()          # mx.metal.clear_cache()
```

Comment on `run_bench` finally says “drop it between tracks.” The `finally` wraps the **whole combo loop**, so weights are dropped once at the end of the matrix, not between tracks.

## Chunking vs peak

The guard already decided 28 rows do not fit and ran **1 row per pass**:

```
Chunking heuristic: 28 rows over 28 passes (rows_per_chunk=1)
```

That limits *concurrent* rows. It does **not** shrink `width × vocab` for a long suffix, and it does not return Metal’s cached buffers to macOS. Budgeting uses `_ASSUMED_BYTES_PER_ROW_SLOPE = 1.0` until a probe succeeds (engine.py SHORTCUT, “can only UNDER-budget”).

## Expected RAM model (for reviewers)

```
Activity Monitor “used”
  ≈ 4 GB weights
  + live logits/KV for THIS case
  + Metal buffer cache ≈ size of the LARGEST slab already allocated
```

Not `4 GB × number of cases`. Not nine models. After a huge case, a tiny case still shows high RAM because bucket 3 (GPU leftover) stays.

## Possible directions (not done here)

1. Call `mx.metal.clear_cache()` on a cadence (per combo / per N cases / when `get_cache_memory()` exceeds a cap). Trade: slower next pass, Activity Monitor honest.
2. Call it from `run_bench` as well as `run_bench_models` (single-model path currently skips it).
3. Align the `run_bench` comment with control flow (“after all combos”, not “between tracks”).
4. Log `mx.metal.get_cache_memory()` / `get_active_memory()` / `get_peak_memory()` per case in the bench log so this is not Activity-Monitor folklore.
5. Decide whether `permutations=rotations` on typesafe is in-scope for a “quality” latency/RAM story (many extra forwards, cache stays hot).

Do **not** unload the 7B after every case (reload 4 GB × hundreds of questions).

## Related finding from the same smoke (separate bug)

Default `--runs 2`. After run 1/2 of `parallel-slots-bundled` wrote `predictions.jsonl`:

```
FAILED combo parallel-slots-bundled (run_failed): RuntimeError:
.../parallel-slots-bundled/predictions.jsonl already exists.
Use --resume to continue, or remove the directory.
```

`run_eval` refuses to overwrite; bench still loops `range(runs)` into the same dir with `resume=_has_manifest` from **before** run 1. Frozen, not patched. Can be split to its own issue if wanted.

## Artifacts

- `#63` comments: 7B-only start, parity + run-2 crash
- Local: `m5-2026-09-19/bench-quality/m5max-128gb-quality/{parity.json,parallel-slots-bundled/,parallel-slots-typesafe/,bench-quality.log}`

**Comment by bnsd55** [comment]
Coordinator: accepted, good diagnosis. Fix scoped as a freeze exception to `jevmlx/bench.py` only: (1) `run_bench` clears the Metal buffer cache after every combo (today only the multi-model path does); (2) per-combo memory telemetry (`mx.metal` peak/active/cache) in the bench log and in `run.json`, peak reset per combo; (3) BENCHMARKING.md paragraph on the RAM model. No model unload between cases, no change to chunking or the slope probe. Direction 5 (rotations on typesafe in the quality run) stays as is for this run; it is the correctness track. The `--runs 2` bug is already fixed in #78 (freeze `be1ec64`). New freeze hash will be posted on #63 when this lands.

**Comment by bnsd55** [comment]
Fixed on main by #80 (freeze `2e94081`): per-combo Metal cache clear in `run_bench`, memory telemetry per combo. Leaving the issue open until the M5 rerun confirms Activity Monitor stays in the expected band.

**Comment by bnsd55** [comment]
Data point after #80 (per-combo clear), 7B smoke on `2e94081`: `[memory] parallel-slots-bundled: peak=4.56 GB active=3.99 GB cache=98.01 GB`. Active memory is small; the Metal allocator hoards ~94 GB of freed buffers within one combo. Next fix: a cache limit via `mx.metal.set_cache_limit` at bench start (default 8 GB, CLI flag), bench.py only.

**Comment by bnsd55** [comment]
Cache cap landed in #81 (`7ccf3ab`): `--metal-cache-gb`, default 8, recorded as `metal_cache_limit_bytes` in run.json. Issue stays open until the M5 full run confirms cache memory stays under the cap.

**Comment by bnsd55** [comment]
Confirmed in the field: the uncapped smoke on `2e94081` fell from ~170 cases/h (run 1) to ~20 cases/h (run 2) with cache=98 GB. Restarting M5 on `7ccf3ab` with the 8 GB cap to measure the fix.
